test/test_nn.py

```diff
@@ -2071,23 +2071,46 @@ def test_embedding_dense_grad_cuda(self):
         self._test_embedding_dense_grad("cuda")
 
     def test_embedding_sparse_backward(self):
+        self._test_embedding_backward(False)
+
+    @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
+    def test_embedding_sparse_half_backward(self):
+        # same as test_embedding_sparse_backward above but testing half types in
+        # cuda. cpu sum not supported for half types.
+        self._test_embedding_backward(True)
+
+    def _test_embedding_backward(self, half_cuda=True):
```

Comment:
I'm a little lost on what `half_cuda` means when it's false.  Does it use cuda but not half?  Does it use half but not cuda?  Does it use neither?

It seems much more natural to just take a dtype and a device?

Reply:
updated to do this